File: 01_Arrays_Hashing/Easy/c#/IsHappy.cs
/*
    ---- Optimal approach space complexity O(1) ---- 
    -> Problem
        -> Determine if a number is a "happy number" using Floyd’s Cycle Detection.
        -> Replace the number with the sum of squares of its digits repeatedly.
        -> If it reaches 1 → happy. If it loops forever → not happy.

    -> Approach (Floyd's Cycle Detection)
        -> No HashSet needed.
        -> Use two pointers:
            - slow moves one step at a time
            - fast moves two steps at a time
        -> If they meet at 1 → happy
           else if they meet elsewhere → not happy

    -> Helper function:
        -> next(n): returns the sum of the squares of the digits of n.

    -> Complexity
        -> Time:  O(log n × k)
        -> Space: O(1)
*/

public class Solution2 {
    // Helper function to compute sum of squares of digits
    private int Next(int n) {
        int sum = 0;
        while (n > 0) {
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }
        return sum;
    }

    public bool IsHappy(int n) {
        int slow = n;
        int fast = Next(n);

        // Move slow by 1 step, fast by 2 steps
        while (fast != 1 && slow != fast) {
            slow = Next(slow);
            fast = Next(Next(fast));
        }

        // If fast reached 1, it’s a happy number
        return fast == 1;
    }
}




/*
    -> Problem
        -> A number is called a "Happy Number" if by repeatedly replacing it 
           with the sum of the squares of its digits, we eventually reach 1.
        -> If we fall into a cycle that does not include 1, the number is not happy.
    
    -> Constraints
        -> 1 <= n <= 2^31 - 1
        -> The number can have multiple digits.

    -> Approach
        -> Use a HashSet<int> to record all numbers that have been seen.
           If a number repeats, we are in a cycle -> return false.
        -> For each step:
            - Split the number into its digits.
            - Compute the sum of squares of all digits.
            - Replace n with this sum.
        -> Continue until:
            - n == 1  -> return true
            - or n is repeated -> return false

    -> Complexity
        -> Time:  O(log n) * k     (each iteration processes all digits; k = number of iterations)
        -> Space: O(k)             (due to HashSet for visited numbers)
*/

public class Solution {
    public bool IsHappy(int n) {
        HashSet<int> seen = new HashSet<int>();
        
        while (n != 1 && !seen.Contains(n)) {
            seen.Add(n);
            int sum = 0;

            while (n > 0) {
                int digit = n % 10;      // extract last digit
                sum += digit * digit;    // add its square
                n /= 10;                 // remove last digit
            }

            n = sum; // assign the new number for the next iteration
        }

        return n == 1;
    }
}

File: 01_Arrays_Hashing/Easy/python/is_happy.py
"""
    -> Problem
        -> Determine if a number is a 'Happy Number' using Floyd’s Cycle Detection.
        -> Replace the number with the sum of the squares of its digits repeatedly.
        -> If it reaches 1 -> happy; if it enters a cycle -> not happy.

    -> Approach (Floyd’s Cycle Detection)
        -> Use two pointers:
            - slow moves one step at a time
            - fast moves two steps at a time
        -> If they meet at 1 -> happy number
        -> If they meet elsewhere -> cycle detected (not happy)
        -> This avoids using extra memory (O(1) space).

    -> Complexity
        -> Time:  O(k × log n) -> O(log n)
        -> Space: O(1)
"""

class Solution:
    def _next(self,n: int) -> int:
        sum = 0
        while n > 0:
            digit = n % 10
            sum += int(digit) ** 2
            n //= 10
        return sum

    def isHappy(self, n: int) -> bool:
        slow = n
        fast = self._next(n)
        while fast != 1 and slow != fast:
            slow = self._next(slow)
            fast = self._next(self._next(fast))
        return fast == 1
        




"""
    -> Problem
        -> A number is called a "Happy Number" if by repeatedly replacing it 
           with the sum of the squares of its digits, we eventually reach 1.
        -> If we fall into a cycle that does not include 1, the number is not happy.
    
    -> Constraints
        -> 1 <= n <= 2^31 - 1
        -> The number can have multiple digits.

    -> Approach
        -> Use a HashSet<int> to record all numbers that have been seen.
           If a number repeats, we are in a cycle -> return false.
        -> For each step:
            - Split the number into its digits.
            - Compute the sum of squares of all digits.
            - Replace n with this sum.
        -> Continue until:
            - n == 1  -> return true
            - or n is repeated -> return false

    -> Time Complexity
        -> Each iteration processes all digits → O(log n)
        -> The sequence length (k) is small and independent of n
           → Total: O(k × log n) -> O(log n)
           (not O(n × log n) because n is a numeric value, not number of items)

    -> Space Complexity
        -> O(k) due to the 'seen' set (bounded, small constant space)
"""
class Solution:
    def isHappy(self, n: int) -> bool:
        seen = set()

        while n != 1 and n not in seen:
            seen.add(n)
            sum = 0
            while n > 0:
                digit = n % 10
                sum += int(digit) ** 2
                n //= 10 
            n = sum
        return n == 1